ldap/data_source_resource_ldap_user.go

```diff
@@ -0,0 +1,80 @@
+package ldap
+
+import (
+	"github.com/Ouest-France/goldap"
+	"github.com/go-ldap/ldap/v3"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+)
+
+func dataSourceLDAPUser() *schema.Resource {
+	return &schema.Resource{
+		Read: dataSourceLDAPUserRead,
+		Schema: map[string]*schema.Schema{
+			"ou": &schema.Schema{
+				Type:     schema.TypeString,
+				Required: true,
+			},
+			"name": &schema.Schema{
+				Type:     schema.TypeString,
+				Optional: true,
+			},
+			"sam_account_name": &schema.Schema{
+				Type:     schema.TypeString,
+				Optional: true,
+			},
+			"user_principal_name": &schema.Schema{
+				Type:     schema.TypeString,
+				Optional: true,
+			},
+			"description": &schema.Schema{
+				Type:     schema.TypeString,
+				Optional: true,
+			},
+		},
+	}
+}
+
+func dataSourceLDAPUserRead(d *schema.ResourceData, m interface{}) error {
+	return resourceLDAPUserRead(d, m)
+}
+
+func resourceLDAPUserRead(d *schema.ResourceData, m interface{}) error {
+	client := m.(*goldap.Client)
+
+	user, err := client.ReadUser(d.Get("ou").(string), d.Get("name").(string), d.Get("sam_account_name").(string), d.Get("user_principal_name").(string))
+	if err != nil {
+		if err.(*ldap.Error).ResultCode == ldap.LDAPResultNoSuchObject {
+			// Object doesn't exist
+			d.SetId("")
+		}
```

Comment:
I'm not sure of the best way to handle that because in my understanding, on a resource when the read method doesn't find the resource we have to remove it from the state:
> When you create something in Terraform but delete it manually, Terraform should gracefully handle it. If the API returns an error when the resource doesn't exist, the read function should check to see if the resource is available first. If the resource isn't available, the function should set the ID to an empty string so Terraform "destroys" the resource in state. The following code snippet is an example of how this can be implemented; you do not need to add this to your configuration for this tutorial.
But in a datasource, when the resource is not found it should throw an error, like in this aws datasource: https://github.com/hashicorp/terraform-provider-aws/blob/main/aws/data_source_aws_ami.go#L244

In the ldap_group datasource I updated the code to handle that with a context value, to know if the call is from a datasource and decide if the code should error or clean the state: https://github.com/Ouest-France/terraform-provider-ldap/blob/master/ldap/resource_ldap_group.go#L82

What's your opinion on this?

Reply:
Or it can be simpler to just duplicate the code in datasource and resource Read function...

Reply:
right. fixed.